docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+- Run _un-ordered_, each generator will see the same input compilation.
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+
+## Implementation
+
+At the simplest level source generators are an implementation of `Microsoft.CodeAnalysis.ISourceGenerator`
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface ISourceGenerator
+    {
+        void Execute(SourceGeneratorContext context);
+    }
+}
+```
+
+Generator implementations are defined in external assemblies passed to the compiler
+using the same `-analyzer:` option used for diagnostic analyzers.
+
+An assembly can contain a mix of diagnostic analyzers and source generators.
+Since generators are loaded from external assemblies, a generator cannot be used to build
+the assembly in which it is defined.
+
+`ISourceGenerator` has a single `Execute` method that is called by the host (either the IDE
+or the command-line compiler). `Execute` passes an instance of `SourceGeneratorContext` that provides access 
+to the `Compilation` and allows the generator to alter it by adding source and reporting diagnostics.
+
+```csharp
+    public struct SourceGeneratorContext
+    {
+        public Compilation Compilation { get; }
+
+        // TODO: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
+
+        public void AddSource(string fileNameHint, SourceText sourceText) { throw new NotImplementedException(); }
+    }
+}
+```
+It is assumed that some generators will want to generate more than one `SourceText`, for example in a 1:1 mapping 
+for additional files. The `fileNameHint` parameter of `AddSource` is intended to address this:
+
+1. If the generated files are emitted to disk, having some ability to put some distinguishing text might be useful. 
+For example, if you have two `.resx` files, generating the files with simply names of `ResxGeneratedFile1.cs` and
+`ResxGeneratedFile2.cs` wouldn't be terribly useful -- you'd want it to be something like
+`ResxGeneratedFile-Strings.cs` and `ResxGeneratedFile-Icons.cs` if you had two `.resx` files
+ named "Strings" and "Icons" respectively.
+
+2. The IDE needs some concept of a "stable" identifier. Source generators create a couple of fun problems for the IDE: 
+users will want to be able to set breakpoints in a generated file, for example. If a source generator outputs multiple 
+files we need to know which is which so we can know which file the breakpoints go with. A source generator of course is 
+allowed to stop emitting a file if needs change (if you delete a `.resx`, then the generated file associated with it 
```

Comment:
> if needs change

It may be better to say "if cancellation is requested" since it has a clearer relation to the API.

Reply:
Slightly different case. Reworded to make clearer.

---
In reply to: [354523960](https://github.com/dotnet/roslyn/pull/40162#discussion_r354523960) [](ancestors = 354523960)